### PR TITLE
Add setter and getter for TCP_QUICKACK on TcpStream for Linux

### PR DESCRIPTION
**Pull request by berendjan**
Reference issue #96256

Setting TCP_QUICKACK on TcpStream for Linux

### CODE REVIEW

**Comment by berendjan** [comment]
r? rust-lang/libs-api @rustbot label +T-libs-api -T-libs

**Comment by the8472** [comment]
Since this is linux-specific I think it should be an extension trait, similar to [os::linux::fs::MetadataExt](https://doc.rust-lang.org/std/os/linux/fs/trait.MetadataExt.html), albeit as a sealed trait and under `os::linux::net` instead.

**Comment by berendjan** [comment]
Agreed, however I tried that approach but ran into the problem that the Socket in TcpStream is private. How do i solve this without creating 2 methods (get/set) for TcpStream?

**Comment by the8472** [comment]
There's the internal trait `sys_common::AsInner` to go from the public to the inner structs.

**Comment by berendjan** [comment]
Ah okay, thank you! Thanks for the tip! Ill update the PR.

On 23 Apr 2022, at 17:52, the8472 ***@***.***> wrote:

﻿
There's the internal trait sys_common::AsInner to go from the public to the inner structs.

**Comment by berendjan** [comment]
@dtolnay hi, I finally got around to updating the PR, please let me know what I can improve!

**Comment by berendjan** [comment]
@the8472 hi, I finally got around to updating the PR, please let me know what I can improve!

File: library/std/src/os/linux/net.rs
Comment (the8472) [inline library/std/src/os/linux/net.rs:53]: Since these tests are `no_run` it would be good to have actual tests in a test module.

File: library/std/src/sys_common/net.rs
Comment (the8472) [inline library/std/src/sys_common/net.rs:340]: The `sys_common` module isn't supposed to contain code behind platform-specifc `cfg`s, instead the `linux::net` module should do enough unwrapping steps to get to the `os::unix` types directly.

File: library/std/src/os/linux/net.rs
Comment (berendjan) [reply library/std/src/os/linux/net.rs:53]: can I also just remove the `no_run`?

Comment (the8472) [reply library/std/src/os/linux/net.rs:53]: I don't know how rustdoc tests interact with cfgs and PR CI only runs linux tests, so it were to fail it would likely only fail after a merge.

But if you have some non-linux machine you could try and see what happens there?

Comment (berendjan) [reply library/std/src/os/linux/net.rs:53]: ah yes it failed already, I found one test for the `nodelay` and adapted it for `quickack`. Hope this works

File: library/std/src/sys_common/net.rs
Comment (berendjan) [reply library/std/src/sys_common/net.rs]: I added a to_inner method to expose the Socket

**Comment by berendjan** [comment]
@the8472 I ran `./x.py check && ./x.py test` without errors on Macbook from 2015 (took some time). Any other checks you recommend?

File: library/std/src/os/linux/tests.rs
Comment (the8472) [inline library/std/src/os/linux/tests.rs:16]: That cfg doesn't make sense, sgx will never show up as linux.

Instead the whole test should simply be cfg'ed to only run on target_os android or linux.

**Review by the8472 — COMMENTED** [review]
I only suggested doing it on a not-linux machine in case you want to remove the `no_run` and see if it impacts another platform. But since you have written a separate test that'll run fine in CI.

For local testing std changes the following two commands should do the job (on a linux machine, since you're making linux-specific changes) :

```
./x.py test library/std --stage 0
./x.py test tidy
```

You'll also want to squash your git history down to one or two commits, 13 is a bit much for a minor addition.

File: library/std/src/os/linux/tests.rs
Comment (berendjan) [reply library/std/src/os/linux/tests.rs]: Yeah about android, I don't know if this option is supported. The official (java) socket API only lists `TCP_NODELAY` as an option. Also don't know how to test this as I don't have an android device.

Comment (the8472) [reply library/std/src/os/linux/tests.rs]: The rust-on-android means native code, not java code. Since setting this just uses the `setsockopt` syscall it has been available on linux kernels since forever, so there shouldn't be a compatibility problem. At worst it'll return an error that the particular socket option isn't supported but that's covered by the `Result` return type.

Comment (berendjan) [reply library/std/src/os/linux/tests.rs]: okay that's clear, ill change it

**Comment by berendjan** [comment]
I'm contemplating closing this PR and open a new one with all the changes in a single commit. I squashed once already and cannot squash again the same squash.

**Comment by the8472** [comment]
you can squash them into a single commit with the `git rebase -i` command

**Comment by berendjan** [comment]
Seems like it worked, thanks for the tip!

**Comment by berendjan** [comment]
> I only suggested doing it on a not-linux machine in case you want to remove the `no_run` and see if it impacts another platform. But since you have written a separate test that'll run fine in CI.
> 
> For local testing std changes the following two commands should do the job (on a linux machine, since you're making linux-specific changes) :
> 
> ```
> ./x.py test library/std --stage 0
> ./x.py test tidy
> ```
> 
> You'll also want to squash your git history down to one or two commits, 13 is a bit much for a minor addition.

I tested both successfully on Debian10 Buster

**Comment by berendjan** [comment]
@the8472 whats next?

File: library/std/src/os/linux/net.rs
Comment (the8472) [inline library/std/src/os/linux/net.rs:22]: This seems copied from the manpage. I don't know if that's ok from the licencing perspective.

**Review by the8472 — COMMENTED** [review]
The tests module isn't declare anywhere, so it hasn't run any tests so far. Try running `x.py test --no-doc --stage 0 library/std --test-args quickack`, that should execute only your (1) test.

Other than that I'm not seeing further issues. I'll leave it to the assigned libs-api team member to decide whether we actually want this API. I recall some conversations about getting `socket2` stuff into std, so this probably qualifies, but quickack is a bit of a niche feature.

File: library/std/src/os/linux/net.rs
Comment (joshtriplett) [reply library/std/src/os/linux/net.rs:22]: It isn't.

Comment (joshtriplett) [inline library/std/src/os/linux/net.rs:22]:
```suggestion
    /// Enable or disable `TCP_QUICKACK`.
    ///
    /// This flag causes Linux to eagerly send ACKs rather than delaying them. Linux may reset this flag after further operations on the socket.
    ///
    /// See [`man 7 tcp`](https://man7.org/linux/man-pages/man7/tcp.7.html) and [TCP delayed acknowledgement](https://en.wikipedia.org/wiki/TCP_delayed_acknowledgment) for more information.
```

(Modulo word-wrapping.)

**Comment by joshtriplett** [comment]
r=me with the comment fixed.

**Comment by berendjan** [comment]
working on the test, it indeed wasn't called. Also since this is an android feature as well, I probably should implement the same methods there as well.

**Comment by berendjan** [comment]
actually, I should move it to `library/std/src/os/unix` I think

**Comment by the8472** [comment]
`os::linux` doesn't cover android? That's... not great. Moving it to `os::unix` won't be appropriate either since it wouldn't work for systems that aren't linux-based.

**Comment by berendjan** [comment]
Well there is a directory `/library/std/src/os/android` so I presume it does not. There are other methods in `/unix/net/addr.rs` that have `#[cfg(any(doc, target_os="android", target_os="linux",))]` so it seems appropriate.

**Comment by the8472** [comment]
> Well there is a directory /library/std/src/os/android so I presume it does not. 

I think it should go there too then, maybe as a reexport?

> There are other methods in /unix/net/addr.rs that have #[cfg(any(doc, target_os="android", target_os="linux",))] so it seems appropriate.

It appears that that is frowned upon (https://github.com/rust-lang/rust/issues/76915#issuecomment-1118979669)

**Comment by berendjan** [comment]
I see, I was considering just adding the module with a cfg, but that is basically kicking the problem down the road until the next `.net` extension. Ill see if I can get the reexport working.

**Comment by berendjan** [comment]
2 questions:
I hardcoded the port number but this is probably  a bad idea since other tests might accidentally use the same port number and this might lead to flaky tests (port being used at same time). `library/std/src/net/test.rs` uses `next_test_ip4()`, but this is not pub and not available in the os mod. What is a good approach?

Also not sure how to properly re-export something, is
```rust
pub use super::linux::net::TcpStreamExt;
```
correct?

**Comment by the8472** [comment]
> Also not sure how to properly re-export something, is
> 
> ```rust
> pub use super::linux::net::TcpStreamExt;
> ```
> 
> correct?

That won't work because the linux module is only conditionally available behind `cfg` flags. Instead you'll need to have this in an internal common module behind a flag for both linux and android and that common module then gets re-exported under linux and android.
The internal `os::fd` module reexported in `unix` and `wasi` would be a precedent.

>  I hardcoded the port number but this is probably a bad idea since other tests might accidentally use the same port number and this might lead to flaky tests (port being used at same time). library/std/src/net/test.rs uses next_test_ip4(), but this is not pub and not available in the os mod. What is a good approach?

I don't know why ` next_test_ip4()` exists, binding to port 0 should let the OS pick one that is free. But you could make `mod test` public to make it accessible.

**Comment by berendjan** [comment]
apparently linux does not like port 0, switching to `next_test_ipv4`

File: library/std/src/os/net/mod.rs
Comment (the8472) [inline library/std/src/os/net/mod.rs:6]: This is `os::net::net`? That leads to confusing re-exports. We need better naming here.

Comment (berendjan) [reply library/std/src/os/net/mod.rs]: agreed, changed to `os::net::tcp`

File: library/std/src/net/mod.rs
Comment (the8472) [inline library/std/src/net/mod.rs:44]: `pub(crate)` should be sufficient here.

File: library/std/src/os/android/net.rs
Comment (the8472) [inline library/std/src/os/android/net.rs:4]: superfluous blank line

Comment (the8472) [inline library/std/src/os/android/net.rs:4]: superfluous blank line

File: library/std/src/os/linux/net.rs
Comment (the8472) [inline library/std/src/os/linux/net.rs:4]: ditto

Comment (the8472) [inline library/std/src/os/linux/net.rs:4]: ditto

File: library/std/src/os/net/mod.rs
Comment (the8472) [inline library/std/src/os/net/mod.rs:5]: ditto

Comment (the8472) [inline library/std/src/os/net/mod.rs:5]: ditto

File: library/std/src/os/linux/mod.rs
Comment (the8472) [inline library/std/src/os/linux/mod.rs:12]: You're still exporting `TcpStreamExt` directly from `os::linux` rather than exporting the `net` module which then contains `TcpStreamExt`.

You might want to build the documentation locally with `x.py doc library/std --open` and check that everything is in the right place.

Comment (the8472) [inline library/std/src/os/linux/mod.rs:12]: You're still exporting `TcpStreamExt` directly from `os::linux` rather than exporting the `net` module which then contains `TcpStreamExt`.

You might want to build the documentation locally with `x.py doc library/std --open` and check that everything is in the right place.

File: library/std/src/net/mod.rs
Comment (berendjan) [reply library/std/src/net/mod.rs]: good one!

File: library/std/src/os/android/net.rs
Comment (berendjan) [reply library/std/src/os/android/net.rs]: I saw this blank line appearing in more files, so I took it as standard. I removed line 4 in all places

File: library/std/src/os/linux/mod.rs
Comment (berendjan) [reply library/std/src/os/linux/mod.rs]: I removed the export and exposed `pub net`

**Comment by berendjan** [comment]
@the8472 thanks for the review, I appreciate it!

**Comment by berendjan** [comment]
@the8472 If you're okay with the PR could you change the label to `S-waiting-for-review`?
After that it's just waiting for someone who can accept or deny the PR?

**Comment by JohnCSimon** [comment]
Triage: 

@berendjan 

FYI: when a PR is ready for review, send a message containing 
`@rustbot ready` to switch to `S-waiting-on-review` so the PR is in the reviewer's backlog.

setting it... 

@rustbot ready

**Review by dtolnay — APPROVED** [review]
Thanks, this is terrific. :)

**Comment by Dylan-DPC** [comment]
failed in [rollup ci](https://github.com/rust-lang/rust/pull/99003#issuecomment-1177186371)

**Comment by berendjan** [comment]
FYI I added `doc` to the cfg include for `net` module (std/src/os/mod.rs) and `AsInner` to imports (std/src/sys_common/net.rs)

**Comment by berendjan** [comment]
@Dylan-DPC I know it's been a while and I should have updated sooner, however is it possible to try the roll up again? I added `doc` to `std/src/os/mod.rs`, I figured that was the problem since it couldn't find during doc generation.

**Comment by Dylan-DPC** [comment]
i think it is better to get a final review from the reviewer first

**Review by dtolnay — APPROVED** [review]

**Comment by berendjan** [comment]
@dtolnay I don't understand the error, what happend?

**Comment by dtolnay** [comment]
Not sure. @bors retry

**Review by joshtriplett — APPROVED** [review]

**Pull request merged**
